### PR TITLE
Move Media field to layouts and use Bootstrap modal and popover

### PR DESCRIPTION
**Pull request by dgrammatiko**
### This is for version 3.5

Moves media field to layouts
Uses bootstrap for the modal
Uses popover for the image preview

Preview:
screen shot 2015-01-22 at 11 58 43
screen shot 2015-01-22 at 11 59 07
#### B/C

None, if we use multiple layouts (e.g. at joomla root one that uses mootools, and on isis and protostar one that uses bootstrap)

Testing

Try to re edit a banner on backend and select an image...

### CODE REVIEW

File: layouts/joomla/form/field/media-js.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * @package    Joomla.Form
+ *
+ * @copyright  Copyright (C) 2005 - 2015 Open Source Matters, Inc. All rights reserved.
+ * @license    GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+// No direct access
+defined('_JEXEC') or die;
+
+extract($displayData);
```
Comment (Fedik) [inline layouts/joomla/form/field/media-js.php]: please add variable descriptions,
```php
/**
 * Layout variables
 * @var string $previewWidth Some description
 * @var string $previewHeight Some description
 */
```
same for other layouts

**Comment by dgt41** [comment]
@Fedik yes you are right, the variable descriptions are missing.
Also creating a media-js.php file is questionable (it doesn’t follow any known pattern)
@okonomiyaki3000 is doing something similar at #5863
So I guess it would be nice if a pattern could be established for layout assets (js, css)

**Comment by Fedik** [comment]
@dgt41 please keep format:
```php
/**
 * @var string $variable1
 * @var string $variable2 
 */
```
as it can be recognized by some IDE

**Comment by dgt41** [comment]
@mbabker Michael you have any solution to this com_media/views mess? What I did, and I don’t like it at all, was to duplicate the views from admin area. I guess there should be a way to use the admin files but also have the ability to override in the from end template.

**Comment by mbabker** [comment]
The issue ultimately lies in [JViewLegacy::_addPath()](https://github.com/joomla/joomla-cms/blob/staging/libraries/legacy/view/legacy.php#L756) and how it adds lookup paths.  Essentially, every additional path that's added to the lookup array gets bumped to the front of the line.  The first way I can think to make this work is that when `$this->getView()` is called in `MediaController::display()`, we'll have to inject the `base_path` option into the config array so that the base lookup path is always `JPATH_COMPONENT_ADMINISTRATOR` (since the site component has no views).  Doing that, the default behavior of setting the template paths should work as it'll first prefer `administrator/components/com_media/views/<whatever>` then fall back to the active template regardless of site or admin.

File: administrator/components/com_media/controller.php [hunk 2]
```diff
@@ -31,25 +31,37 @@ public function display($cachable = false, $urlparams = false)
...
+
 		switch ($vName)
 		{
 			case 'images':
 				$vLayout = $this->input->get('layout', 'default', 'string');
 				$mName = 'manager';
-
+				// Get/Create the view
+				$view = $this->getView($vName, $vType);
+				if (JFactory::getApplication()->isSite() && !file_exists(JPATH_ROOT . '/templates/' . JFactory::getApplication()->getTemplate() .
```
Comment (dgt41) [inline administrator/components/com_media/controller.php]: @mbabker This works for me, with or without template override, frontend or backend. But I end up refactoring a little bit this part of the controller. Is that ok?

Comment (mbabker) [inline administrator/components/com_media/controller.php]: Do we really need to move the getView calls?  In theory only the two views are usable from the front-end, but should something be refactored to make the media manager more usable in that context, we're just going to end up moving things around again.  I'd also really suggest not making this coupled to the existence of a single file; what I suggested with injecting the path into the getView call should suffice for all cases without extra conditionals.

Comment (dgt41) [inline administrator/components/com_media/controller.php]: `$view->set('_basePath', $this->basePath);` will do the trick but fronted overrides are not working, and the point here is to have a template override coupled with a layout so designers can do their thing!

Comment (mbabker) [inline administrator/components/com_media/controller.php]: My main point is that you don't need to move the view instantiation into the switch conditional or add this if conditional if you inject the base path when the view is created in the original spot as I suggested.  JViewLegacy will always prefer the active template override then fall back to the base path if no other paths are specified.  Forcing the base path to `JPATH_COMPONENT_ADMINISTRATOR` via the parameter injection ensures that com_media's admin layouts are ALWAYS the last path lookup with the active template (regardless of site or admin request) taking a higher precedence in the path lookup.

Comment (dgt41) [inline administrator/components/com_media/controller.php]: @mbabker You were absolutely right! Thanks!

**Comment by designbengel** [comment]
Modals are cut off and not responsive. See screenshots for details.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5871">issues.joomla.org/joomla-cms/5871</a>.</sub>

**Comment by designbengel** [comment]
screen shot 2015-03-14 at 08 08 16screen shot 2015-03-14 at 08 08 19<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5871">issues.joomla.org/joomla-cms/5871</a>.</sub>

**Comment by dgt41** [comment]
@designbengel I am not getting this and frankly this cannot happen as the modal width is set as a percentage of the window with (80%). mot probably your emulator is messing up there

**Comment by adhocgraFX** [comment]
Parse error: syntax error, unexpected end of file in C:\xampp\htdocs\test\administrator\templates\isis\html\layouts\joomla\form\field\media.php on line 138<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5871">issues.joomla.org/joomla-cms/5871</a>.</sub>

**Comment by dgt41** [comment]
@adhocgraFX 
Please try reapplying the patch as line 138 is just pure html `</div>`
Also can you provide some more infos how to replicate this?

**Comment by adhocgraFX** [comment]
@dgt41 got same error on another pc/ xamp installation / same php settings: display errors = On, error reporting = E ALL | E STRICT, safe mode = Off, magic quotes gpc = Off<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5871">issues.joomla.org/joomla-cms/5871</a>.</sub>

**Comment by dgt41** [comment]
@adhocgraFX what’s in line 138 of C:\xampp\htdocs\test\administrator\templates\isis\html\layouts\joomla\form\field\media.php
?
Suppose that this is where the error is thrown..

**Comment by dgt41** [comment]
@adhocgraFX And also which view triggered this error?

**Comment by adhocgraFX** [comment]
e.g. in http://localhost/joomla-cms/administrator/index.php?option=com_banners&view=banner&layout=edit&id=2 
but in more tested views like com contact and so on
media php, line 138: </div>
i changed now in line 125 and 128 <? endif; ?> to <?php endif; ?>
and now it works > 
i'm not a pro-programmer, but i think, this was the issue, better ask @mbabker too

**Comment by dgt41** [comment]
@adhocgraFX I must admit I was looking at the wrong file 🙈 Thanks for finding this, should be fixed now

**Comment by N6REJ** [comment]
I thought we weren't supposed to use "short codes" in J!?
Bear

**Comment by mbabker** [comment]
We don't use them in Joomla core and IIRC in PHP 5.3 short PHP codes are something that can be turned off.

**Comment by dgt41** [comment]
@N6REJ @mbabker Actually I never intended to use short codes, that was a typo

**Comment by n9iels** [comment]
Good move! I have test this PR using the select image field in <i>Global configuration</i> and noticed a few things:

- On the latest version of Google Chrome, I see two scrollbars inside the modal.
image-modal
- On iPhone 5 the select image field and the model himself are fall outside the viewport.
- Very small detail: the preview button only has rounded corners when hover.
- When scrolling inside the modal, and you reach the end, the rest of the page is scrolling farther. Not sure if this is specifiek for your PR, or a bug in all Bootstrap modals.

Maybe you can add the #6462 (when that PR is merged)? That will brings the buttons on a more logical place in my opinion.

**Comment by dgt41** [comment]
@n9iels thanks for testing
most of the bugs you mention here are caused because of a change in modals height in 3.4.1 which obviously has to be taken into consideration here! (the height of the iframe is way to big that’s the reason for the blank space and the double scrollbars)
The idea for the buttons at the bottom is a nice one and I have to implemented here and also to other modals (we need some consistency)
preview button: thats a minor css thing
On iPhone 5 the select image field and the model: again css thing

BUT this awaits the new layouts renderer, which is still not coded, so actually, even if it seems ready, it has to wait for 3.5

**Comment by Harmageddon** [comment]
I just tested this to see if it fixes #6799. In Isis, the image preview works fine. The path tooltips however only seem to work when I select a new image after applying the patch. For images which were set before applying the patch, I don't get tooltips. Tested in the "images and links" section of the article form.
As you already remarked, @dgt41, the behavior of this patch is not consistent in Isis and Hathor. Not sure if it is related to this problem, but I don't see a good reason for creating overrides for Isis, our default admin template.

**Comment by dgt41** [comment]
@Harmageddon The tooltip should be fine for stored values as well now!
About the different layout in Hathor this was intended so people can test both layouts (mootools and bootstrap). I guess when the layout renderer is ready and all the tests are good here copying the overrides to hathor shouldn’t be a problem.
@n9iels I made some changes:
buttons should be fine now,
iPhone should have no problems.
Footer is on it’s way (on another commit 😃)

**Comment by n9iels** [comment]
@test Stil found some view-poort bug for iPhone:
media modal iphone

On desktop, the "Choose directory" is cut of, and still have 2 scroll bars:
media modal

On your screenshot, this issn't the matter. Maybe a different display size? I'm using a 23 insh monitor

**Comment by dgt41** [comment]
@n9iels the message "Geen bestand…" which is running out of screen is a browser specific (check the code) thus it might seem broken in emulator but in iOS should render fine.
This a real rendering with IOS 8.3 and iPhone 6
img_0065

But I wouldn’t care too much for the inner part of the modal, as with the new media manager this will change 😉

**Comment by n9iels** [comment]
didn't knew that, thanks for the information.
I hope the new media manger will come soon :)

**Comment by dgt41** [comment]
@mbabker this doesn’t look good: `PHP Fatal error:  require(): Cannot redeclare class php_codecoverage_report_xml_tests in /home/travis/.phpenv/versions/5.3.29/bin/phpunit on line 208`

**Comment by mbabker** [comment]
See https://www.traviscistatus.com/incidents/h91j8ft0qs2l

**Comment by dgt41** [comment]
The last iteration restores the scrolling on iOS devices and sets a more proper height for the iframe
img_0135

**Comment by anibalsanchez** [comment]
#Test OK

One step closer to drop mootools. Congratulations!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5871">issues.joomla.org/joomla-cms/5871</a>.</sub>

**Comment by Kubik-Rubik** [comment]
Restarted Travis process.

**Comment by frogydiak** [comment]
Sorry to mix in with different perspective but have anyone noticed that it is nothing but an uploader? It should not be called Media Manager at all. IMO, the popup should have Create Folder, Delete Folder, Move files, Copy files functionalities within the popup.

**Comment by dgt41** [comment]
@frogydiak There is a PR for a new Media manager #3839 you can take this idea there

**Comment by dgt41** [comment]
@Fedik To relief some of the pain with the reputable I end up using exactly the same function as the old mootools code, and make some other changes so the values are injected automagically from the calling element. So in theory you only have to replace the id’s on the span for the preview, on the input field and on the remove button (as well as the actual modal id). And then override the modal toggler. Should be a bit easier. I hope so!

**Comment by Fedik** [comment]
@dgt41  thank you for the work! 
hope I will find some time this week to check it more

**Comment by Fedik** [comment]
it close to what I thought but still hardly linked to ID, 
please check alternative https://github.com/dgt41/joomla-cms/pull/22 :wink:

**Comment by dgt41** [comment]
What has changed since the last RTC:
- Dropped the inline script by utilizing data attributes for passing the variables from PHP to JS
- Introduced mediafield.js and mediafield-mootools.js as well as their minified versions
- Introduce minified version for the rest of the media scripts
- Changes in the HTML structure followed by changes in the javascript to ensure that the field will be compatible with the repeatable fields
- Bring the tabs from #3839 thanks @Buddhima
- Drop the media/images/default.php override in Isis and Protostar (don't ask)
- Include the recommendations from @phproberto to utilize a function getLayoutData()

Please test again, to get this ready for 3.5

**Comment by zero-24** [comment]
I have tested this item :white_check_mark: successfully on 5838d664f48b2c47c5590aa7c73d9500684e1fc8<br /><br />Works good. I like the improvment e.g. the new "Tab" for uploading. Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5871">issues.joomla.org/joomla-cms/5871</a>.</sub>

**Comment by designbengel** [comment]
There are two scrollbars in the modal that might cause issues on touch devises?
bildschirmfoto 2015-11-01 um 17 57 53

**Comment by dgt41** [comment]
@designbengel no problem for the mobiles (or smaller screen devices)
 
img_0236

I made some minor touches so another test is needed...

**Comment by zero-24** [comment]
@dgt41 this is how it looks on chrome now: 
media_field_chrome

**Comment by dgt41** [comment]
@zero-24 now all browsers are looking good ;)
joomla-git - administration - articles edit 2015-11-03 18-17-10

File: administrator/components/com_media/views/imageslist/tmpl/default.php [hunk 3]
```diff
@@ -19,6 +19,16 @@
 }
 
 JFactory::getDocument()->addScriptDeclaration("var ImageManager = window.parent.ImageManager;");
+Jfactory::getDocument()->addStyleDeclaration(
```
Comment (zero-24) [inline administrator/components/com_media/views/imageslist/tmpl/default.php]: Should be ` JFactory` i guess ;)

**Comment by zero-24** [comment]
Ok looks good now to me. Just on very minimal code comment ;) :+1:

**Comment by zero-24** [comment]
I have tested this item :white_check_mark: successfully on 3934118d1db6ab2bf8660e73a0a988f56b7605d5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5871">issues.joomla.org/joomla-cms/5871</a>.</sub>

File: administrator/components/com_media/views/imageslist/tmpl/default.php (hunk 3 above)
Comment (dgt41) [inline administrator/components/com_media/views/imageslist/tmpl/default.php]: 👍

**Comment by zero-24** [comment]
I have tested this item :white_check_mark: successfully on ab0da6d146a9cdfbf9232bbedb25d7021882451c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5871">issues.joomla.org/joomla-cms/5871</a>.</sub>

**Comment by dgt41** [comment]
In case anyone tries to test this you need to apply also #8248

**Comment by phproberto** [comment]
I see the scroll bar overlapped here too.

media-modal

**Comment by dgt41** [comment]
@phproberto can you try emptying browsers cache?

**Comment by phproberto** [comment]
I have tested this item :white_check_mark: successfully on f5df125874ace5a0bc4056a3a2745b1eacb789ab<br /><br />Everything works now. Thanks!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5871">issues.joomla.org/joomla-cms/5871</a>.</sub>

**Comment by designbengel** [comment]
I have tested this item :white_check_mark: successfully on f5df125874ace5a0bc4056a3a2745b1eacb789ab<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5871">issues.joomla.org/joomla-cms/5871</a>.</sub>

**Comment by zero-24** [comment]
RTC Thanks :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5871">issues.joomla.org/joomla-cms/5871</a>.</sub>

**Comment by infograf768** [comment]
This has created a regression
See: https://github.com/joomla/joomla-cms/issues/8367

**Pull request merged**